pkg/queue/bounded_queue_test.go

```diff
@@ -156,3 +157,173 @@ func (s *consumerState) assertConsumed(expected map[string]bool) {
 	}
 	assert.Equal(s.t, expected, s.snapshot())
 }
+
+func TestResizeUp(t *testing.T) {
+	q := NewBoundedQueue(2, func(item interface{}) {
+		fmt.Printf("dropped: %v\n", item)
+	})
+
+	var firstConsumer, secondConsumer, releaseConsumers sync.WaitGroup
+	firstConsumer.Add(1)
+	secondConsumer.Add(1)
+	releaseConsumers.Add(1)
+
+	released, resized := false, false
+	q.StartConsumers(1, func(item interface{}) {
+		if !resized { // we'll have a second consumer once the queue is resized
+			// signal that the worker is processing
+			firstConsumer.Done()
+		} else {
+			// once we release the lock, we might end up with multiple calls to reach this
+			if !released {
+				secondConsumer.Done()
+			}
+		}
+
+		// wait until we are signaled that we can finish
+		releaseConsumers.Wait()
+	})
+
+	assert.True(t, q.Produce("a")) // in process
+	firstConsumer.Wait()
+
+	assert.True(t, q.Produce("b"))  // in queue
+	assert.True(t, q.Produce("c"))  // in queue
+	assert.False(t, q.Produce("d")) // dropped
+	assert.EqualValues(t, 2, q.Capacity())
+	assert.EqualValues(t, q.Capacity(), q.Size())
+	assert.EqualValues(t, q.Capacity(), len(*q.items))
+
+	resized = true
+	assert.True(t, q.Resize(4))
+	assert.True(t, q.Produce("e")) // in process by the second consumer
+	secondConsumer.Wait()
+
+	assert.True(t, q.Produce("f"))  // in the new queue
+	assert.True(t, q.Produce("g"))  // in the new queue
+	assert.False(t, q.Produce("h")) // the new queue has the capacity, but the sum of queues doesn't
+
+	assert.EqualValues(t, 4, q.Capacity())
+	assert.EqualValues(t, q.Capacity(), q.Size()) // the combined queues are at the capacity right now
+	assert.EqualValues(t, 2, len(*q.items))       // the new internal queue should have two items only
+
+	released = true
+	releaseConsumers.Done()
+}
+
+func TestResizeDown(t *testing.T) {
+	q := NewBoundedQueue(4, func(item interface{}) {
+		fmt.Printf("dropped: %v\n", item)
+	})
+
+	var consumer, releaseConsumers sync.WaitGroup
+	consumer.Add(1)
+	releaseConsumers.Add(1)
+
+	released := false
+	q.StartConsumers(1, func(item interface{}) {
+		// once we release the lock, we might end up with multiple calls to reach this
+		if !released {
```

Comment:
Released does not seem to be needed

Reply:
Why's that? Note that there's quite a bunch of go funcs just waiting for this lock to be released to start calling consumers with other items. If we have *not* released, we mark one step down in the work group. If we have released the consumer, we do not want to count down further, as we'd reach a negative count.

Reply:
I mean you can remove this parameter altogether. It's not changed during the test - only at the end which does not affect tests and the default value satisfy this if statement.

Reply:
Removal of this variable does not have any effect on the tests. It seems redundant

Reply:
It needs to be there, otherwise the test will intermittently fail due to "negative counts" in the work group.